pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/InstanceResponseBlock.java

```diff
@@ -18,26 +18,132 @@
+ * The {@code InstanceResponseBlock} is the holder of the server side results.
  */
 public class InstanceResponseBlock implements Block {
-  private final DataTable _instanceResponseDataTable;
+  private final BaseResultsBlock _resultsBlock;
+  private final QueryContext _queryContext;
+  private final Map<Integer, String> _exceptions;
+  private final Map<String, String> _metadata;
 
-  public InstanceResponseBlock(DataTable dataTable) {
-    _instanceResponseDataTable = dataTable;
+  public InstanceResponseBlock(BaseResultsBlock resultsBlock, QueryContext queryContext) {
+    _resultsBlock = resultsBlock;
+    _queryContext = queryContext;
+    _exceptions = new HashMap<>();
+    List<ProcessingException> processingExceptions = resultsBlock.getProcessingExceptions();
+    if (processingExceptions != null) {
+      for (ProcessingException processingException : processingExceptions) {
+        _exceptions.put(processingException.getErrorCode(), processingException.getMessage());
+      }
+    }
+    _metadata = resultsBlock.getResultsMetadata();
   }
 
-  public DataTable getInstanceResponseDataTable() {
-    return _instanceResponseDataTable;
+  /**
+   * Metadata only instance response.
+   */
+  public InstanceResponseBlock() {
+    _resultsBlock = null;
+    _queryContext = null;
+    _exceptions = new HashMap<>();
+    _metadata = new HashMap<>();
+  }
+
+  private InstanceResponseBlock(Map<Integer, String> exceptions, Map<String, String> metadata) {
+    _resultsBlock = null;
+    _queryContext = null;
+    _exceptions = exceptions;
+    _metadata = metadata;
+  }
+
+  public InstanceResponseBlock toMetadataOnlyResponseBlock() {
+    return new InstanceResponseBlock(_exceptions, _metadata);
+  }
+
+  public void addException(ProcessingException processingException) {
+    _exceptions.put(processingException.getErrorCode(), processingException.getMessage());
+  }
+
+  public void addException(int errorCode, String exceptionMessage) {
+    _exceptions.put(errorCode, exceptionMessage);
+  }
+
+  public void addMetadata(String key, String value) {
+    _metadata.put(key, value);
+  }
+
+  @Nullable
+  public BaseResultsBlock getResultsBlock() {
+    return _resultsBlock;
+  }
+
+  @Nullable
+  public QueryContext getQueryContext() {
+    return _queryContext;
+  }
+
+  public Map<Integer, String> getExceptions() {
+    return _exceptions;
+  }
+
+  public Map<String, String> getResponseMetadata() {
+    return _metadata;
+  }
+
+  @Nullable
+  public DataSchema getDataSchema() {
+    return _resultsBlock != null ? _resultsBlock.getDataSchema(_queryContext) : null;
+  }
+
+  @Nullable
+  public Collection<Object[]> getRows() {
+    return _resultsBlock != null ? _resultsBlock.getRows(_queryContext) : null;
+  }
+
+  public DataTable toDataTable()
+      throws IOException {
+    DataTable dataTable = toDataOnlyDataTable();
+    attachMetadata(dataTable);
+    return dataTable;
+  }
```

Comment:
is there any situation one will use the BaseResultsBlock API rather than the ones in InstanceResponseBlock?

if not can we make all the abstract methods in BaseResultsBlock package private and move the InstanceResponseBlock into the results package?

Reply:
Currently no, but I feel we may leave it public in case we need it in the future. The `Block` interface is not for public usage, so there shouldn't be compatible issue.